File: internal/core/completion_test.go
package core

import (
	"strings"
	"testing"
)

// The whole point of the completion gate: 30 VU where only 10 requests finished
// measured nothing about capacity, so it must never be reported as stable.
func TestRunWithMostRequestsUnfinishedIsNotStable(t *testing.T) {
	run := Run{
		Config: RunConfig{Mode: LoadModeVU, VUs: 30, MaxErrorPercent: 2, MaxP95Millis: 60000, MinCompletionPercent: 95},
		Result: RunResult{
			Successes: 10, Total: 10, Issued: 30, Completed: 10, Cancelled: 20,
			CompletionPercent: 33.3, Latency: Distribution{P95Millis: 5000},
		},
	}
	if IsRunStable(run) {
		t.Fatal("a run that finished a third of its requests was judged stable")
	}
	message := InstabilityMessage(run)
	if !strings.Contains(message, "33.3%") || !strings.Contains(message, "30 started") {
		t.Fatalf("instability message hides the shortfall: %q", message)
	}
}

func TestFullyCompletedRunWithinSLOStaysStable(t *testing.T) {
	run := Run{
		Config: RunConfig{Mode: LoadModeVU, VUs: 30, MaxErrorPercent: 2, MaxP95Millis: 60000, MinCompletionPercent: 95},
		Result: RunResult{
			Successes: 30, Total: 30, Issued: 30, Completed: 30,
			CompletionPercent: 100, Latency: Distribution{P95Millis: 5000},
		},
	}
	if !IsRunStable(run) {
		t.Fatalf("fully completed run within SLO judged unstable: %s", InstabilityMessage(run))
	}
}

// Results from an Agent that predates the lifecycle counters must not be gated
// on a completion rate it never reported.
func TestResultWithoutLifecycleCountersIsNotGated(t *testing.T) {
	run := Run{
		Config: RunConfig{Mode: LoadModeVU, VUs: 5, MaxErrorPercent: 2, MaxP95Millis: 2000},
		Result: RunResult{Successes: 10, Total: 10, Latency: Distribution{P95Millis: 100}},
	}
	if _, _, short := CompletionShortfall(run); short {
		t.Fatal("gated a result that reports no request lifecycle")
	}
	if !IsRunStable(run) {
		t.Fatal("legacy result judged unstable by the completion gate")
	}
}

// The auto search must not raise the load off an unfinished run.
func TestAutoSearchStopsWhenTheStartingLoadDidNotComplete(t *testing.T) {
	search := AutoSearch{Status: AutoSearchRunning, Config: AutoSearchConfig{MaxLoad: 40, Run: RunConfig{Mode: LoadModeVU}}}
	run := Run{
		Config: RunConfig{Mode: LoadModeVU, VUs: 10, MaxErrorPercent: 2, MaxP95Millis: 60000, MinCompletionPercent: 95},
		Result: RunResult{Successes: 2, Total: 2, Issued: 10, Completed: 2, Cancelled: 8, CompletionPercent: 20, Latency: Distribution{P95Millis: 1000}},
	}
	next, more := AdvanceSearch(&search, run)
	if more || next != 0 {
		t.Fatalf("search escalated the load off an unfinished run: next=%d", next)
	}
	if search.RecommendedLoad != 0 {
		t.Fatalf("unfinished run produced a recommended capacity: %d", search.RecommendedLoad)
	}
	if !strings.Contains(search.Message, "finished") {
		t.Fatalf("search message does not explain the shortfall: %q", search.Message)
	}
}

func TestSteadyStateWindowMustBeShorterThanTheRun(t *testing.T) {
	base := RunConfig{Mode: LoadModeVU, VUs: 1, DurationSeconds: 60, MaxTokens: 1024}
	valid := base
	valid.SteadyStateSeconds = 30
	if err := ValidateRunConfig(valid); err != nil {
		t.Fatalf("valid steady window rejected: %v", err)
	}
	invalid := base
	invalid.SteadyStateSeconds = 60
	if err := ValidateRunConfig(invalid); err == nil {
		t.Fatal("steady window equal to the run duration was accepted, leaving no samples")
	}
}

File: web/test/completion.test.js
import assert from "node:assert/strict";
import test from "node:test";
import { completionShortfall, getVerdict } from "../src/results.js";
import { operationalSummary } from "../src/operational-summary.js";

const partialRun = {
  status: "completed",
  config: { mode: "vu", vus: 30, max_error_percent: 2, max_p95_millis: 600000, min_completion_percent: 95 },
  result: {
    issued: 30, completed: 10, cancelled: 20, completion_percent: 33.3,
    successes: 10, failures: 0, total: 10, latency: { p95_millis: 5000 },
  },
};

test("a run that finished a third of its requests reports the shortfall", () => {
  const shortfall = completionShortfall(partialRun);
  assert.ok(shortfall, "expected a shortfall");
  assert.equal(shortfall.issued, 30);
  assert.equal(shortfall.completed, 10);
  assert.match(shortfall.message, /30건 중 10건/);
  assert.match(shortfall.message, /20건은 측정 종료 시점에 아직 진행 중/);
});

test("the shortfall outranks the latency verdict and blocks a stable judgement", () => {
  const verdict = getVerdict(partialRun);
  assert.equal(verdict.status, "at-risk");
  assert.match(verdict.recommendation, /안정 용량으로 판정하지 않습니다/);
});

test("the operational summary refuses to claim the load was served", () => {
  const summary = operationalSummary(partialRun);
  assert.equal(summary.status, "at-risk");
  assert.equal(summary.cause, "incomplete");
  assert.match(summary.nextAction, /30 VU 수용 가능을 증명하지 못합니다/);
});

test("a fully completed run within SLO is still judged stable", () => {
  const run = {
    status: "completed",
    config: { mode: "vu", vus: 30, max_error_percent: 2, max_p95_millis: 600000, min_completion_percent: 95 },
    result: { issued: 30, completed: 30, cancelled: 0, completion_percent: 100, successes: 30, failures: 0, total: 30, latency: { p95_millis: 5000 } },
  };
  assert.equal(completionShortfall(run), null);
  assert.equal(getVerdict(run).status, "stable");
  assert.equal(operationalSummary(run).status, "stable");
});

test("a result without lifecycle counters is not gated on completion", () => {
  const legacy = {
    status: "completed",
    config: { mode: "vu", vus: 5, max_error_percent: 2, max_p95_millis: 2000 },
    result: { successes: 10, failures: 0, total: 10, latency: { p95_millis: 100 } },
  };
  assert.equal(completionShortfall(legacy), null);
  assert.equal(getVerdict(legacy).status, "stable");
});
